src/util/BlockingCache.hpp

```diff
@@ -0,0 +1,221 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2025, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include "util/Assert.hpp"
+#include "util/Mutex.hpp"
+
+#include <boost/asio/error.hpp>
+#include <boost/asio/spawn.hpp>
+#include <boost/asio/steady_timer.hpp>
+#include <boost/signals2/connection.hpp>
+#include <boost/signals2/signal.hpp>
+#include <boost/signals2/variadic_signal.hpp>
+
+#include <atomic>
+#include <concepts>
+#include <expected>
+#include <functional>
+#include <optional>
+#include <shared_mutex>
+#include <utility>
+
+namespace util {
+
+/**
+ * @brief A thread-safe cache that blocks getting operations until the cache is updated
+ *
+ * @tparam ValueType The type of value to be cached
+ * @tparam ErrorType The type of error that can occur during updates
+ */
+template <typename ValueType, typename ErrorType>
+    requires(not std::same_as<ValueType, ErrorType>)
+class BlockingCache {
+public:
+    /**
+     * @brief Possible states of the cache
+     */
+    enum class State { Empty, Updating, Full };
```

Comment:
Perhaps `Empty` and `Full` are not the right names for the actual states they represent?
`Empty` seems to mean `NotCached` and `Full` seems to mean `Cached` or `Available` if i'm reading the code correctly. Wdyt?

Reply:
I meant to read it as "cache is empty" or "cache is full". I think `Cached` or `NotCached` is rather related to an entry but not to cache.